### PR TITLE
Use externs files when producing docs

### LINKED ISSUE #3503: Require existence of externs files when producing docs
This originally arose in my mind as a way of fixing #3264, but it would be a breaking change to the CLI so I think a dedicated issue is probably warranted.

At the moment, we invoke the type checker during generation of docs in order to fill in any type signatures for declarations where the type signatures have not already been provided. However, we can't perform all of the desugaring steps which would normally have been run by this point, because some of them modify the AST in such a way that we wouldn't be able to produce sensible docs from them (e.g. desugaring of type classes and type synonyms). Often this is not a problem, but occasionally we see errors in docs generation which only occur when type signatures are left off (e.g. #3264).

To address this, I am proposing that we instead modify `purs docs` to accept an `--output` option, like `purs compile` does, and have `purs docs` first check that up-to-date externs files exist for all of the provided source files, and fail if not. Then, `purs docs` could obtain any missing types from the externs files. As with `purs compile`, the `--output` option would have a default of "output". However, it would only be used for locating externs files.

A nice side effect of doing this would be that it would prevent situations where the docs have incorrect types. For instance, if I write

```purescript
foo :: String
foo = 3
```

in a module and run `purs docs` on it, provided that all value declarations in that module have type signatures, it will succeed and the resulting docs will say that `foo` has the type `String`.

This would also pave the way for using the re-exports information contained in externs files (see #2191) to replace most of the code in `Docs.Convert.ReExports`. I would be very happy about this, because not only is that code quite frankly horrible (I'm allowed to say that because I wrote it), but also I suspect it's one of the main causes of the awful performance and memory consumption of `purs docs`. However, to do that, I think we'd need to add the module a re-export is being re-exported from to the externs format (currently, we only include the module a re-export originally came from, see https://github.com/purescript/purescript/pull/2191#issuecomment-225186677).

### LINKED ISSUE #3624: `purs docs` fails due to thinking a Newtype instance is using a type synonym
Apologies if this is a duplicate.

Originally reported JordanMartinez/learn-halogen#5 and later in slamdata/purescript-halogen#601, running `spago docs` on the `purescript-halogen`'s `v5.0.0-rc.4` release will fail with the below error. [In his comment in the Halogen issue](https://github.com/slamdata/purescript-halogen/issues/601#issuecomment-489799441), Gary said this is likely a bug in `purs docs`.

```
Generating documentation for the project. This might take a while..
Error found:
in module Halogen.Query.EventSource
at .spago/halogen/v5.0.0-rc.4/src/Halogen/Query/EventSource.purs:32:8 - 32:66 (line 32, column 8 - line 32, column 66)

  Type class instances for type synonyms are disallowed.

in type class instance
                                                         
  Data.Newtype.Newtype (EventSource m a)                 
                       (m                                
                          { finalizer :: Finalizer m     
                          , producer :: Producer a m Unit
                          }                              
                       )                                 
                                                         

See https://github.com/purescript/documentation/blob/master/errors/TypeSynonymInstance.md for more information,
or to contribute content related to this error.


spago: Docs generation failed.
```

### PR DESCRIPTION
**Pull request by hdgarrood**
Resolves #3503, fixes #3624.

This commit does a few related things:

**Docs codegen target**

Adds a `docs` codegen target for `purs compile`. When supplied, a
`docs.json` file is written for each module alongside the other files in
that module's directory within the output directory. The file contains a
JSON-serialized Docs.Types.Module, which is the same as the one which
will be produced by `purs publish` for that module, except that the
re-exports field will be empty; this is because computing re-exports
requires access to documentation for all of the upstream modules, and we
don't have an easy way of getting hold of all of these (which won't
adversely affect memory consumption / performance for the regular
compilation path).

Since we are producing docs during `purs compile`, we now have easy
access to the relevant externs files, which we can use to fill in type
signatures of any declarations which don't already have their own. This
means that no module within the `Language.PureScript.Docs` namespace
needs to call into the type checker any more in order to get accurate
and useful type signatures in docs, regardless of whether annotations
have been given or not, which both speeds up docs generation
significantly and eradicates the possibility of bugs like #3624 arising.

**Use docs.json files in purs docs and purs publish**

The subcommands `purs docs` and `purs publish` now each accept an
`--output` option, which should point to a compiler output directory,
whose default value is `"output"`. With this information, we can reach
into the compiler output directory to grab the information in each of
the required `docs.json` files, perform one final pass over them to fill
in re-exports, and then produce documentation in whatever format is
requested. If the `docs.json` files are not up-to-date with respect to
the source files which have been provided to `purs docs`, we throw an
error.

In effect, `purs docs` and `purs publish` now behave more like
`purs bundle` does, in the sense of getting their input data from the
output of `purs compile`, rather than from a collection of source files.

**Refactor Docs library API**

The Docs library API has been refactored and simplified; now, the `purs
docs` CLI and the `Language.PureScript.Publish` module only really need
to use one function for getting hold of docs: the function `collectDocs`
from the module `Language.PureScript.Docs.Collect`. This function
implements the behaviour described above for reading `docs.json` files,
making sure they are up-to-date, and filling in re-exports.

With this new `collectDocs` function, a number of other functions no
longer need to be exposed as part of the compiler library API. In
particular, the module `Language.PureScript.Docs.ParseInPackage` has
been removed entirely.

### CODE REVIEW

File: src/Language/PureScript/Errors.hs [hunk 1]
```diff
@@ -1060,6 +1061,12 @@ prettyPrintSingleError (PPEOptions codeColor full level showDocs relPath) e = fl
             , indent $ paras $ map (line . markCode . showQualified showOp) (NEL.toList ops)
             , line "Use parentheses to resolve this ambiguity."
             ]
+    renderSimpleErrorMessage NeedToBuildDocs =
+      paras
+        [ line "Some of the input modules are missing up-to-date documentation data in the compiler output directory. Please compile all of the relevant modules using `purs compile' with the `docs' codegen target."
+        , line "For example:"
+        , indent $ line "$ purs compile --codegen js,docs SOURCE_FILES"
```
Comment (natefaubion) [inline src/Language/PureScript/Errors.hs]: Is the `js` option necessary? I don't think it is, and I'm sure the example is meant to indicate "hey, you can compile docs with normal JS anytime you want", but it makes it seem likes it's necessary.

Comment (hdgarrood) [reply src/Language/PureScript/Errors.hs]: You're right, it isn't necessary. Do you think we should just write `--codegen docs` here then? I don't have any strong feelings either way.

File: src/Language/PureScript/ModuleDependencies.hs [hunk 2]
```diff
@@ -50,8 +50,7 @@ sortModules toSig ms = do
     return (fst <$> ms'', moduleGraph)
   where
     toGraphNode :: S.Set ModuleName -> (a, ModuleSignature) -> m ((a, ModuleSignature), ModuleName, [ModuleName])
-    toGraphNode mns m@(_, ModuleSignature _ mn ds) = do
-      let deps = ordNub (mapMaybe usedModules ds)
```
Comment (natefaubion) [inline src/Language/PureScript/ModuleDependencies.hs]: What's the reason for removing the ordNub? Was it unnecessary?

Comment (hdgarrood) [reply src/Language/PureScript/ModuleDependencies.hs]: Oops, no, that was an oversight. I'll put it back.

File: src/Language/PureScript/Errors.hs (hunk 1 above)
Comment (natefaubion) [reply src/Language/PureScript/Errors.hs]: I don't think it's necessary. The help for `purs compile` indicates that you can provide multiple codegen targets.

File: src/Language/PureScript/ModuleDependencies.hs (hunk 2 above)
Comment (natefaubion) [reply src/Language/PureScript/ModuleDependencies.hs]: I mean, I'm genuinely unsure what it's doing, so I was hoping you know 😆.

Comment (hdgarrood) [reply src/Language/PureScript/ModuleDependencies.hs]: Actually, maybe it would make sense to use a `Set` for the `sigImports` field?

Comment (hdgarrood) [reply src/Language/PureScript/ModuleDependencies.hs]: I think it is actually necessary if we want the module graph to not contain duplicate entries (which indeed seems like a sensible thing to want).

Comment (natefaubion) [reply src/Language/PureScript/ModuleDependencies.hs]: Does the specific order matter though. `Set` wouldn't preserve that.

Comment (hdgarrood) [reply src/Language/PureScript/ModuleDependencies.hs]: That's true. I'm not sure. Perhaps best to put the `ordNub` back to leave it as it was then.

**Comment by hdgarrood** [comment]
Actually, perhaps the CLI flag should be `--compile-output`, so as not to clash with an `--output` flag which might be used to specify the file or directory you'd like the result to be written to.

**Comment by natefaubion** [comment]
I think `--compile-output` makes sense.

**Comment by hdgarrood** [comment]
Sorry for changing this at the last minute, but I think it makes the CLI a lot nicer. Instead of erroring if the docs are not up-to-date, we just build them. This makes this a non-breaking change. Now that we are actually building, might want to consider adding a `--json-errors` flag to `purs docs` too, but that can happen later.

**Comment by hdgarrood** [comment]
Thank you both!

**Pull request merged**